--- minimal	2020-06-11 09:29:42.070331633 +0200
+++ ../../../boot.l	2020-06-11 09:28:43.454054994 +0200
@@ -1,8 +1,10 @@
 (define list (lambda args args))
 
 (define %print print)
 (define %dump  dump)
 
+(define error)
+
 (define %error
   (lambda args
     (set error abort)
@@ -11,15 +15,15 @@
     (%print "\n")
     (abort)))
 
-(define error
+(set error
   (lambda args
     (set error %error)
-    (%print "\nerror: ")
+    (%print "\n[31;1merror: ")
     (apply print args)
-    (%print "\n")
+    (%print "[m\n")
     (abort)))
 
 (define caar (lambda (x) (car (car x))))
 (define cadr (lambda (x) (car (cdr x))))
 (define cdar (lambda (x) (cdr (car x))))
 (define cddr (lambda (x) (cdr (cdr x))))
